# Data Analysis & Manipulation with Pandas

In this notebook, I will focus on importing datasets, dealing with missing values, changing data types, filtering, sorting, selecting specific column(s), dealing with duplicate values, dropping and adding rows and columns, counting values, counting unique values.

During my graduation project, I have dealt with a lot of Data Pre-processing stuff without using any libraries. I have just used if statements and for loops. It literally killed me. After I started to learn Pandas, I decided to write this article to help students and beginners to get start with it. I will do my best to introduce you with Pandas’ some of the most useful capabilities in the stage of Exploratory Data Analysis. 

We are going to use the famous Titanic Dataset which is available on Kaggle.

Let’s start with importing Pandas.

In [1]:
import pandas as pd

The common shortcut of Pandas is pd. Instead of writing “pandas.” we can write “pd.” now. So, there is a dot after “pd” which is used to call a method from Pandas library.

## Importing the dataset with read_csv

To read a dataset, we are going to use read_csv.

In [3]:
df = pd.read_csv("train.csv")

## Checking first elements of the DataFrame with .head() method

After we run the code above, nothing will appear. So you have to write df to see your data. But instead of seeing all the data, we are going to use the “.head()” method to see the first five elements of the data. Before you run the read_csv code, you can write df.head() below. So it’s going to be like this:

In [4]:
df = pd.read_csv("train.csv")
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Inside the parentheses, we can write the number of elements that we want to see. If we leave it blank, it will show the first five elements. If we write 25 inside of the parentheses, it will show the first 25 elements of the dataframe.

In [5]:
df.head(25)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   
20           21         0       2   
21           22         1       2   
22           23         1       3   
23           24         1       1   
24           25         0       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   
20                               Fynney, Mr. Joseph J    male  35.0      0   
21                              Beesley, Mr. Lawrence    male  34.0      0   
22                        McGowan, Miss. Anna "Annie"  female  15.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   
24                      Palsson, Miss. Torborg Danira  female   8.0      3   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
5       0            330877   8.4583   NaN        Q  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S  
11      0            113783  26.5500  C103        S  
12      0         A/5. 2151   8.0500   NaN        S  
13      5            347082  31.2750   NaN        S  
14      0            350406   7.8542   NaN        S  
15      0            248706  16.0000   NaN        S  
16      1            382652  29.1250   NaN        Q  
17      0            244373  13.0000   N

## Checking last elements of the DataFrame with .tail() method

There is also a method to see the see last n number of elements. The method is called .tail().

The same rule is also applied here. If we leave the parentheses blank, it will be set as 5, if we write 25 inside of the parentheses, it will show the last 25 elements of the dataframe.

In [6]:
df.tail(25)

PassengerId  Survived  Pclass  \
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                 Name     Sex   Age  SibSp  \
866                      Duran y More, Miss. Asuncion  female  27.0      1   
867              Roebling, Mr. Washington Augustus II    male  31.0      0   
868                       van Melkebeke, Mr. Philemon    male   NaN      0   
869                   Johnson, Master. Harold Theodor    male   4.0      1   
870                                 Balkic, Mr. Cerin    male  26.0      0   
871  Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                          Carlsson, Mr. Frans Olof    male  33.0      0   
873                       Vander Cruyssen, Mr. Victor    male  47.0      0   
874             Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0      1   
875                  Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
876                     Gustafsson, Mr. Alfred Ossian    male  20.0      0   
877                              Petroff, Mr. Nedelio    male  19.0      0   
878                                Laleff, Mr. Kristo    male   NaN      0   
879     Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880      Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
881                                Markun, Mr. Johann    male  33.0      0   
882                      Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
883                     Banfield, Mr. Frederick James    male  28.0      0   
884                            Sutehall, Mr. Henry Jr    male  25.0      0   
885              Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                             Montvila, Rev. Juozas    male  27.0      0   
887                      Graham, Miss. Margaret Edith  female  19.0      0   
888          Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                             Behr, Mr. Karl Howell    male  26.0      0   
890                               Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare        Cabin Embarked  
866      0     SC/PARIS 2149  13.8583          NaN        C  
867      0          PC 17590  50.4958          A24        S  
868      0            345777   9.5000          NaN        S  
869      1            347742  11.1333          NaN        S  
870      0            349248   7.8958          NaN        S  
871      1             11751  52.5542          D35        S  
872      0               695   5.0000  B51 B53 B55        S  
873      0            345765   9.0000          NaN        S  
874      0         P/PP 3381  24.0000          NaN        C  
875      0              2667   7.2250          NaN        C  
876      0              7534   9.8458          NaN        S  
877      0            349212   7.8958          NaN        S  
878      0            349217   7.8958          NaN        S  
879      1             11767  83.1583          C50        C  
880      1            230433  26.0000          NaN       

### Summary of .head() and .tail()

To sum up, these methods return the top and bottom of the dataframe. The default number of rows is set to 5. But, you can change it by writing number of rows that you want to see inside the parentheses.

## Reading a CSV file with specific columns

Let’s assume that you don’t want all of the columns in your CSV file. You just want 3 of them and you want to get rid of them at the begining. How to do that? Good news! There is a really easy way to do that. We are going to use usecols argument to specify the column names that we want to work with. Let’s work with just PassengerId, Survived, Pclass columns.

In [7]:
df1 = pd.read_csv("train.csv", usecols= ["PassengerId", "Survived", "Pclass"]) 
df1.head() #and add this to see if our code worked

PassengerId  Survived  Pclass
0            1         0       3
1            2         1       1
2            3         1       3
3            4         1       1
4            5         0       3

## Getting some information about dataset with .describe() and .info()

After we load our dataset with read_csv, we would like to get some information about the columns. To do that, we are going to use .describe() and .info()

### .describe() method

This method is used to get a summary of numeric values in your dataset. It calculates the mean, standard deviation, minimum value, maximum value, 1st percentile, 2nd percentile, 3rd percentile of the columns with numeric values. It also counts the number of variables in the dataset. So, we will be able to see if there are missing values in columns.

In [8]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

We can see that count of Age column is 714, mean is 29.6, standard deviation is 14.52 and so on. Thanks to count, we can understand that there are some missing values in this column. We will deal with them later.

### .info() method

This method prints information about a DataFrame including the index dtype and column dtypes, non-null values and memory usage.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


#### We can also see the data types of columns with .dtypes

In [11]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

# Sorting the DataFrame based on a column

## Sorting numeric data

There is a column in our dataframe which represents the price of the ticket that passengers bought. Let’s assume that we want to see lowest ticket price. To do that, we have to use .sort_values() method. It doesn’t matter if the column that you want to sort is string or numeric. If it includes letters, it will sort in alphabetical order.

In [13]:
df.sort_values("Fare").head(10)

PassengerId  Survived  Pclass                             Name   Sex  \
271          272         1       3     Tornquist, Mr. William Henry  male   
597          598         0       3              Johnson, Mr. Alfred  male   
302          303         0       3  Johnson, Mr. William Cahoone Jr  male   
633          634         0       1    Parr, Mr. William Henry Marsh  male   
277          278         0       2      Parkes, Mr. Francis "Frank"  male   
413          414         0       2   Cunningham, Mr. Alfred Fleming  male   
674          675         0       2       Watson, Mr. Ennis Hastings  male   
263          264         0       1            Harrison, Mr. William  male   
466          467         0       2            Campbell, Mr. William  male   
732          733         0       2             Knight, Mr. Robert J  male   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
271  25.0      0      0    LINE   0.0   NaN        S  
597  49.0      0      0    LINE   0.0   NaN        S  
302  19.0      0      0    LINE   0.0   NaN        S  
633   NaN      0      0  112052   0.0   NaN        S  
277   NaN      0      0  239853   0.0   NaN        S  
413   NaN      0      0  239853   0.0   NaN        S  
674   NaN      0      0  239856   0.0   NaN        S  
263  40.0      0      0  112059   0.0   B94        S  
466   NaN      0      0  239853   0.0   NaN        S  
732   NaN      0      0  239855   0.0   NaN        S

We used .head() method to see only first 10 of the results after sorting. We can see that the lowest “Fare” values are 0. What if we want to see the highest fare? All we have to do is use the .tail() method. No, just kidding. We have to set the ascending argument as False. But using .tail() is an alternative :P.

In [14]:
df.sort_values("Fare", ascending = False).head(10)

PassengerId  Survived  Pclass                                   Name  \
258          259         1       1                       Ward, Miss. Anna   
737          738         1       1                 Lesurer, Mr. Gustave J   
679          680         1       1     Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1             Fortune, Miss. Mabel Helen   
27            28         0       1         Fortune, Mr. Charles Alexander   
341          342         1       1         Fortune, Miss. Alice Elizabeth   
438          439         0       1                      Fortune, Mr. Mark   
311          312         1       1             Ryerson, Miss. Emily Borie   
742          743         1       1  Ryerson, Miss. Susan Parker "Suzette"   
118          119         0       1               Baxter, Mr. Quigg Edmond   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292              NaN        C  
737    male  35.0      0      0  PC 17755  512.3292             B101        C  
679    male  36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
88   female  23.0      3      2     19950  263.0000      C23 C25 C27        S  
27     male  19.0      3      2     19950  263.0000      C23 C25 C27        S  
341  female  24.0      3      2     19950  263.0000      C23 C25 C27        S  
438    male  64.0      1      4     19950  263.0000      C23 C25 C27        S  
311  female  18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
742  female  21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
118    male  24.0      0      1  PC 17558  247.5208          B58 B60        C

Thanks to that, we can see the highest fares paid by passengers. If we want to save the sorted version of the dataframe, there are two alternatives. One is the old way, which is

Thanks to that, we can see the highest fares paid by passengers. If we want to save the sorted version of the dataframe, there are two alternatives. One is the old way, which is

In [15]:
df1 = df.sort_values("Fare", ascending = False)

The other way is using the inplace argument. If we set this argument as True, it will write over it.

In [17]:
df1.sort_values("Fare", ascending = False, inplace = True)
df1.head()

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1          Fortune, Miss. Mabel Helen   
341          342         1       1      Fortune, Miss. Alice Elizabeth   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
88   female  23.0      3      2     19950  263.0000  C23 C25 C27        S  
341  female  24.0      3      2     19950  263.0000  C23 C25 C27        S

## Sorting columns with strings

We are going to sort the “Cabin” column. There are lots of missing (NaN) values in this column. How can we deal with them? Thankfully, there is an argument which is called na_position which helps us to set a position for the NaN values in the dataset.

In [19]:
df.sort_values("Cabin", ascending = True, na_position ='last').tail(10)

PassengerId  Survived  Pclass  \
878          879         0       3   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
888          889         0       3   
890          891         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
878                            Laleff, Mr. Kristo    male   NaN      0      0   
880  Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0      1   
881                            Markun, Mr. Johann    male  33.0      0      0   
882                  Dahlberg, Miss. Gerda Ulrika  female  22.0      0      0   
883                 Banfield, Mr. Frederick James    male  28.0      0      0   
884                        Sutehall, Mr. Henry Jr    male  25.0      0      0   
885          Rice, Mrs. William (Margaret Norton)  female  39.0      0      5   
886                         Montvila, Rev. Juozas    male  27.0      0      0   
888      Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1      2   
890                           Dooley, Mr. Patrick    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  
878            349217   7.8958   NaN        S  
880            230433  26.0000   NaN        S  
881            349257   7.8958   NaN        S  
882              7552  10.5167   NaN        S  
883  C.A./SOTON 34068  10.5000   NaN        S  
884   SOTON/OQ 392076   7.0500   NaN        S  
885            382652  29.1250   NaN        Q  
886            211536  13.0000   NaN        S  
888        W./C. 6607  23.4500   NaN        S  
890            370376   7.7500   NaN        Q

By using .tail(10), we can see the last 10 elements of the dataset. If we look at Cabin column, we can see that all of the values are NaN.

To sum up, we learned to read csv files with .read_csv method(with and without selecting specific columns), used .head() and .tail() to see elements in the top and at the bottom, got information about dataset with .describe() and .info(), sorted columns which include string or numeric values (with and without NaN values)

## Counting the occurences of variables

Knowing how many unique variables are there in a column, or the occurence of each item in a column might be very useful in some cases. Let’s count the number of male and female passengers with .value_counts()

### Using .value_counts() to count the occurences of each variable in a column

To count the occurence of a variable, we have to select the column first. You can select a column with two different ways:

In [20]:
df["Sex"].head()

0      male
1    female
2    female
3    female
4      male
Name: Sex, dtype: object

In [21]:
df.Sex.head()

0      male
1    female
2    female
3    female
4      male
Name: Sex, dtype: object

Since .value_counts() is a method, all we have to do is appending this method to the code above. It will look like this:

In [23]:
df["Sex"].value_counts()

male      577
female    314
Name: Sex, dtype: int64

### Using .nunique() to count number of unique values that occur in dataset or in a column

If we want to see number of unique records in a dataset or in a column, we have to use .nunique() method.

In [24]:
df.nunique() #by typing this, we can see the counts of unique numbers in each column

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

We can also count the unique records with .nunique() for a column. All we have to do is add the column name.

In [25]:
df["Embarked"].nunique()

3

If you want to see number of unique records for more than one column, you have to add one more square bracket.

In [26]:
df[["Embarked", "Sex"]].nunique()

Embarked    3
Sex         2
dtype: int64

# Changing the data type

We checked the data types of the columns in Titanic dataset. We saw that the type of Embarked column is object. After counting the unique values in Embarked column with .unique(), we can see that there are 3 unique values in that column. So we can consider that the data type should be categorical. To change the datatype of that column the code below must be executed:

In [27]:
df["Embarked"] = df["Embarked"].astype("category")
df["Embarked"].dtype

CategoricalDtype(categories=['C', 'Q', 'S'], ordered=False)

# Filtering

## Filtering under one condition

The comparison sign in Python is == (double equal sign). So you should double check whether you used 2 equal signs. If you use just one equal sign, you might ruin your data. Let’s assume that I want to see if the “Embarked” column is equal to “C”. The true version of the comparison is:

In [28]:
df["Embarked"] == "C" 

0      False
1       True
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9       True
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19      True
20     False
21     False
22     False
23     False
24     False
25     False
26      True
27     False
28     False
29     False
       ...  
861    False
862    False
863    False
864    False
865    False
866     True
867    False
868    False
869    False
870    False
871    False
872    False
873    False
874     True
875     True
876    False
877    False
878    False
879     True
880    False
881    False
882    False
883    False
884    False
885    False
886    False
887    False
888    False
889     True
890    False
Name: Embarked, Length: 891, dtype: bool

What if we dont want to see just Trues and Falses? What if we want to see all information of those whose Embarked is C? To do that:

In [29]:
df[df["Embarked"] == "C"]

PassengerId  Survived  Pclass  \
1              2         1       1   
9             10         1       2   
19            20         1       3   
26            27         0       3   
30            31         0       1   
31            32         1       1   
34            35         0       1   
36            37         1       3   
39            40         1       3   
42            43         0       3   
43            44         1       2   
48            49         0       3   
52            53         1       1   
54            55         0       1   
57            58         0       3   
60            61         0       3   
64            65         0       1   
65            66         1       3   
73            74         0       3   
96            97         0       1   
97            98         1       1   
111          112         0       3   
114          115         0       3   
118          119         0       1   
122          123         0       2   
125          126         1       3   
128          129         1       3   
130          131         0       3   
135          136         0       2   
139          140         0       1   
..           ...       ...     ...   
716          717         1       1   
731          732         0       3   
737          738         1       1   
742          743         1       1   
762          763         1       3   
766          767         0       1   
773          774         0       3   
780          781         1       3   
789          790         0       1   
793          794         0       1   
798          799         0       3   
803          804         1       3   
817          818         0       2   
827          828         1       2   
830          831         1       3   
832          833         0       3   
835          836         1       1   
839          840         1       1   
842          843         1       1   
843          844         0       3   
847          848         0       3   
849          850         1       1   
852          853         0       3   
858          859         1       3   
859          860         0       3   
866          867         1       2   
874          875         1       2   
875          876         1       3   
879          880         1       1   
889          890         1       1   

                                                  Name     Sex    Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.00      1   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.00      1   
19                             Masselmani, Mrs. Fatima  female    NaN      0   
26                             Emir, Mr. Farred Chehab    male    NaN      0   
30                            Uruchurtu, Don. Manuel E    male  40.00      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female    NaN      1   
34                             Meyer, Mr. Edgar Joseph    male  28.00      1   
36                                    Mamee, Mr. Hanna    male    NaN      0   
39                         Nicola-Yarred, Miss. Jamila  female  14.00      1   
42                                 Kraeff, Mr. Theodor    male    NaN      0   
43            Laroche, Miss. Simonne Marie Anne Andree  female   3.00      1   
48                                 Samaan, Mr. Youssef    male    NaN      2   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.00      1   
54                      Ostby, Mr. Engelhart Cornelius    male  65.00      0   
57                                 Novel, Mr. Mansouer    male  28.50      0   
60                               Sirayanian, Mr. Orsen    male  22.00      0   
64                               Stewart, Mr. Albert A    male    NaN      0   
65                            Moubarek, Master. Gerios    male    NaN      1   
73                         Chronopoulos, Mr. Apostolos    male  26.00      1   
96                           Goldschmidt, Mr. Ge

If we write it like that, pandas will understand that we want to see those rows that have True value.

Another way to do that might be:

In [30]:
embarked_c_mask = df["Embarked"] == "C"
df[embarked_c_mask]

PassengerId  Survived  Pclass  \
1              2         1       1   
9             10         1       2   
19            20         1       3   
26            27         0       3   
30            31         0       1   
31            32         1       1   
34            35         0       1   
36            37         1       3   
39            40         1       3   
42            43         0       3   
43            44         1       2   
48            49         0       3   
52            53         1       1   
54            55         0       1   
57            58         0       3   
60            61         0       3   
64            65         0       1   
65            66         1       3   
73            74         0       3   
96            97         0       1   
97            98         1       1   
111          112         0       3   
114          115         0       3   
118          119         0       1   
122          123         0       2   
125          126         1       3   
128          129         1       3   
130          131         0       3   
135          136         0       2   
139          140         0       1   
..           ...       ...     ...   
716          717         1       1   
731          732         0       3   
737          738         1       1   
742          743         1       1   
762          763         1       3   
766          767         0       1   
773          774         0       3   
780          781         1       3   
789          790         0       1   
793          794         0       1   
798          799         0       3   
803          804         1       3   
817          818         0       2   
827          828         1       2   
830          831         1       3   
832          833         0       3   
835          836         1       1   
839          840         1       1   
842          843         1       1   
843          844         0       3   
847          848         0       3   
849          850         1       1   
852          853         0       3   
858          859         1       3   
859          860         0       3   
866          867         1       2   
874          875         1       2   
875          876         1       3   
879          880         1       1   
889          890         1       1   

                                                  Name     Sex    Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.00      1   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.00      1   
19                             Masselmani, Mrs. Fatima  female    NaN      0   
26                             Emir, Mr. Farred Chehab    male    NaN      0   
30                            Uruchurtu, Don. Manuel E    male  40.00      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female    NaN      1   
34                             Meyer, Mr. Edgar Joseph    male  28.00      1   
36                                    Mamee, Mr. Hanna    male    NaN      0   
39                         Nicola-Yarred, Miss. Jamila  female  14.00      1   
42                                 Kraeff, Mr. Theodor    male    NaN      0   
43            Laroche, Miss. Simonne Marie Anne Andree  female   3.00      1   
48                                 Samaan, Mr. Youssef    male    NaN      2   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.00      1   
54                      Ostby, Mr. Engelhart Cornelius    male  65.00      0   
57                                 Novel, Mr. Mansouer    male  28.50      0   
60                               Sirayanian, Mr. Orsen    male  22.00      0   
64                               Stewart, Mr. Albert A    male    NaN      0   
65                            Moubarek, Master. Gerios    male    NaN      1   
73                         Chronopoulos, Mr. Apostolos    male  26.00      1   
96                           Goldschmidt, Mr. Ge

If we want to filter our data in vice versa, it is going to show the rows that their Embarked column is not “C”.

In [33]:
df[df["Embarked"] != "C"]

PassengerId  Survived  Pclass  \
0              1         0       3   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
32            33         1       3   
33            34         0       2   
35            36         0       1   
37            38         0       3   
..           ...       ...     ...   
854          855         0       2   
855          856         1       3   
856          857         1       1   
857          858         1       1   
860          861         0       3   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
20                                Fynney, Mr. Joseph J    male  35.0      0   
21                               Beesley, Mr. Lawrence    male  34.0      0   
22                         McGowan, Miss. Anna "Annie"  female  15.0

## Filtering under two or more condition

### AND operator

We are going to use AND and OR operator to filter with more than one condition. Let’s assume that we want to see the passengers whose Fare is smaller than 100 and who are female. We are going to create 2 new masks to complete that.

In [34]:
df_fare_mask = df["Fare"] < 100
df_sex_mask = df["Sex"] == "female"
df[df_fare_mask & df_sex_mask]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
14            15         0       3   
15            16         1       2   
18            19         0       3   
19            20         1       3   
22            23         1       3   
24            25         0       3   
25            26         1       3   
28            29         1       3   
32            33         1       3   
38            39         0       3   
39            40         1       3   
40            41         0       3   
41            42         0       2   
43            44         1       2   
44            45         1       3   
47            48         1       3   
49            50         0       3   
52            53         1       1   
53            54         1       2   
56            57         1       2   
58            59         1       2   
61            62         1       1   
66            67         1       2   
..           ...       ...     ...   
801          802         1       2   
807          808         0       3   
809          810         1       1   
813          814         0       3   
816          817         0       3   
820          821         1       1   
823          824         1       3   
829          830         1       1   
830          831         1       3   
835          836         1       1   
842          843         1       1   
849          850         1       1   
852          853         0       3   
853          854         1       1   
854          855         0       2   
855          856         1       3   
858          859         1       3   
862          863         1       1   
863          864         0       3   
865          866         1       2   
866          867         1       2   
871          872         1       1   
874          875         1       2   
875          876         1       3   
879          880         1       1   
880          881         1       2   
882          883         0       3   
885          886         0       3   
887          888         1       1   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN      0   
22                         McGowan, Miss. Anna "Annie"  female  15.0      0   
24                       Palsson, Miss. Torborg Danira  female   8.0      3   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
28                       O'Dwyer, Miss. Ellen "Nellie"  female   NaN      0   
32                            Glynn, Miss. Mary Agatha  female   NaN      0   
38                  Vander Planke, Miss. Augusta Maria  female  18.0      2   
39                         Nicola-Yarred, Miss. Jamila  female  14.0      1   
40      Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.0      1   
41   Turpin, Mrs. William John Robert (Dorothy Ann ...  female  27.0

### OR operator

Let’s do another example with OR operator. We are going to use | sign to do that. Let’s see the passengers whose fare is more than 500 or older than 70.

In [36]:
df_fare_mask2 = df["Fare"] > 500
df_age_mask = df["Age"] > 70
df[df_fare_mask2 | df_age_mask]

PassengerId  Survived  Pclass                                  Name  \
96            97         0       1             Goldschmidt, Mr. George B   
116          117         0       3                  Connors, Mr. Patrick   
258          259         1       1                      Ward, Miss. Anna   
493          494         0       1               Artagaveytia, Mr. Ramon   
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   
679          680         1       1    Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1                Lesurer, Mr. Gustave J   
851          852         0       3                   Svensson, Mr. Johan   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
96     male  71.0      0      0  PC 17754   34.6542           A5        C  
116    male  70.5      0      0    370369    7.7500          NaN        Q  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
493    male  71.0      0      0  PC 17609   49.5042          NaN        C  
630    male  80.0      0      0     27042   30.0000          A23        S  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
851    male  74.0      0      0    347060    7.7750          NaN        S

# Finding the null values with .isnull()

One of the most common problems in data science is missing values. To detect them, there is a beautiful method which is called .isnull(). With this method, we can get a boolean series (True or False). As we did before, by masking the condition, we can extract the values which are null. For example

In [37]:
null_mask = df["Cabin"].isnull()
df[null_mask]

PassengerId  Survived  Pclass  \
0              1         0       3   
2              3         1       3   
4              5         0       3   
5              6         0       3   
7              8         0       3   
8              9         1       3   
9             10         1       2   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
22            23         1       3   
24            25         0       3   
25            26         1       3   
26            27         0       3   
28            29         1       3   
29            30         0       3   
30            31         0       1   
32            33         1       3   
33            34         0       2   
34            35         0       1   
35            36         0       1   
36            37         1       3   
37            38         0       3   
38            39         0       3   
..           ...       ...     ...   
852          853         0       3   
854          855         0       2   
855          856         1       3   
856          857         1       1   
858          859         1       3   
859          860         0       3   
860          861         0       3   
861          862         0       2   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
868          869         0       3   
869          870         1       3   
870          871         0       3   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
888          889         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN      0   
20                                Fynney, Mr. Joseph J    male  35.0      0   
22                         McGowan, Miss. Anna "Annie"  female  15.0      0   
24                       Palsson, Miss. Torborg Danira  female   8.0      3   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
26                             Emir, Mr. Farred Chehab    male   NaN

With this code, we are saying that “Show me the passengers whose cabin is unknown”. The output is:

Instead of using this method on a column, it can be used on whole dataset too. If we want to count of the null values of all columns in a dataframe, we just have to write code below

In [38]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Dealing with missing values

There are lots of dealing ways with missing values but in this article, we are going to use “ignore the tuple” and “fill it with median”. We are going to ignore the “Cabin” column since %70 of that column is missing. And we are going to fill the missing Ages with median value of that column.

### Dropping a column

To drop the “Cabin” column, we have to execute the code below.

In [39]:
df.drop(labels = ["Cabin"], axis=1).head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

We used .drop method to drop Cabin column. There is 2 arguments above. In the labels argument, we have to specify the column names that we want to drop, in the axis argument, we specified that we drop it column-wise.

If you want to drop more than one column, all you have to do is add it in the square brackets. For example:

In [40]:
df.drop(labels = ["Cabin", "Name"], axis=1).head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0            1         0       3    male  22.0      1      0   
1            2         1       1  female  38.0      1      0   
2            3         1       3  female  26.0      0      0   
3            4         1       1  female  35.0      1      0   
4            5         0       3    male  35.0      0      0   

             Ticket     Fare Embarked  
0         A/5 21171   7.2500        S  
1          PC 17599  71.2833        C  
2  STON/O2. 3101282   7.9250        S  
3            113803  53.1000        S  
4            373450   8.0500        S

We are dropping cabin column and name column at the same time. As I mentioned before, If we knew that we won’t use these columns, we would have usecols argument of .read_csv method to get rid of that columns at the beginning.

## Filling missing values with .fillna()

To fill missing values in a dataframe, there is a method called .fillna().

Let’s assume that we have lots of missing values in a column and we want to fill them with 0. All we have to do is write the code below

In [ ]:
df["columnname"].fillna(0, inplace = True) #with inplace argument, we don't have to write it as
df["columnname"] = df["columnname"].fillna(0, inplace = True)

We can fill it with a specific value too. Let’s give an example on Titanic dataset

In [41]:
df["Age"].fillna("Unknown", inplace = True)

But instead of filling with “Unknown” to fill the missing values in the Age column, we are going to use the median of that column. To do that:

In [43]:
df = pd.read_csv("train.csv")

In [44]:
df['Age'] = df['Age'].fillna((df['Age'].median()))

If we want to fill the missing values with mean or something else, all we have to do is change the method at the end. After executing this code, we should check if there is still null values in the Age column.